server/src/main/java/org/elasticsearch/tasks/CancellableTask.java

```diff
@@ -67,6 +70,16 @@ public final String getReasonCancelled() {
         return reason;
     }
 
+    /**
+     * This method registers a listener that needs to be notified when this task is cancelled given that the task is not cancelled yet.
+     */
+    public final boolean registerListener(CancellationListener listener) {
+        if (isCancelled) {
+            return false;
+        }
+        return listeners.add(listener);
```

Comment:
I think I would expect registering a listener on a cancelled task to complete the listener straight away, similarly to how things like `CompletableFuture#whenComplete` and `ListenableActionFuture#addListener` work. It's probably ok as it's used here but I worry this will trap a future developer.

It's possible that the task is cancelled after we check the `isCancelled` flag but before the listener is added, which would mean this method returns `true` but the listener is never completed.

Reply:
You are right, the way it's set up now it's error prone. I tried to preemptively end but it's not worth it